src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt

```diff
@@ -0,0 +1,137 @@
+
+            // don't include hidden index
+            val isHidden = IndexMetadata.INDEX_HIDDEN_SETTING.get(indexMetadata.settings)
+            if (isHidden) return null
+
+            // only process indices created after template
+            // traverse all ism templates for matching ones
+            val patternMatchPredicate = { pattern: String -> Regex.simpleMatch(pattern, indexName) }
+            var matchedPolicy: String? = null
+            var highestPriority: Int = -1
+            ismTemplates.filter { (_, template) ->
+                template.lastUpdatedTime.toEpochMilli() < indexMetadata.creationDate
+            }.forEach { (policyID, template) ->
+                val matched = template.indexPatterns.stream().anyMatch(patternMatchPredicate)
+                if (matched && highestPriority < template.priority) {
+                    highestPriority = template.priority
+                    matchedPolicy = policyID
+                }
+            }
+
+            return matchedPolicy
+        }
+
+        /**
+         * validate the template Name and indexPattern provided in the template
+         * reusing ES validate function in MetadataIndexTemplateService
+         */
+        @Suppress("ComplexMethod")
+        fun validateFormat(indexPatterns: List<String>): ElasticsearchException? {
+            val validationErrors = mutableListOf<String>()
+            for (indexPattern in indexPatterns) {
+                if (indexPattern.contains(" ")) {
+                    validationErrors.add("index_patterns [$indexPattern] must not contain a space")
+                }
+                if (indexPattern.contains(",")) {
+                    validationErrors.add("index_pattern [$indexPattern] must not contain a ','")
+                }
+                if (indexPattern.contains("#")) {
+                    validationErrors.add("index_pattern [$indexPattern] must not contain a '#'")
+                }
+                if (indexPattern.contains(":")) {
+                    validationErrors.add("index_pattern [$indexPattern] must not contain a ':'")
+                }
+                if (indexPattern.startsWith("_")) {
+                    validationErrors.add("index_pattern [$indexPattern] must not start with '_'")
+                }
+                if (!Strings.validFileNameExcludingAstrix(indexPattern)) {
+                    validationErrors.add("index_pattern [" + indexPattern + "] must not contain the following characters " +
+                            Strings.INVALID_FILENAME_CHARS)
+                }
+            }
+
+            if (validationErrors.size > 0) {
+                val validationException = ValidationException()
+                validationException.addValidationErrors(validationErrors)
+                return IndexManagementException.wrap(validationException)
+            }
+            return null
+        }
+
+        /**
+         * find policy templates whose index patterns overlap with given template
```

Comment:
What does overlap mean here? Just based on the index pattern?

Reply:
changed to `conflictingPolicyTemplates`
yeah, it checks if the new template index pattern matching range overlaps with any old template index pattern

Reply:
Hmm is the behavior correct? I see it used below to reject a PutPolicy request, so if there is overlap in index patterns it rejects it? I thought we allow overlap in index patterns, which is why we have the priority to control which one takes priority?

Reply:
overlap index pattern is allowed when priority is different, if priority is same, putPolicy is blocked